app/helpers/automated_tests_helper.rb

```diff
@@ -380,6 +474,435 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+        raise "list_run_scripts is not sorted"
+      end
+      ctr = ctr + 1
+    end
+
+    return list_run_scripts
+  end
+
+  # Request an automated test. Ask Resque to enqueue a job.
+  def self.async_test_request(grouping_id, call_on)
+    if files_available?
+      if has_permission?
+        Resque.enqueue(AutomatedTestsHelper, grouping_id, call_on)
+      end
+    end
+  end
+
+
+  # Verify that MarkUs has some files to run the test.
+  # Note: this does not guarantee all required files are presented.
+  # Instead, it checks if there is at least one test script and
+  # source files are successfully exported.
+  def self.files_available?()
+    test_dir = File.join(MarkusConfigurator.markus_config_automated_tests_repository, @assignment.short_identifier)
+    src_dir = @repo_dir
+    assign_dir = @repo_dir + "/" + @assignment.repository_folder
+    if !(File.exists?(test_dir))
+      # TODO: show the error to user instead of raising a runtime error
+      raise I18n.t("automated_tests.test_files_unavailable")
+    elsif !(File.exists?(src_dir))
+      # TODO: show the error to user instead of raising a runtime error
+      raise I18n.t("automated_tests.source_files_unavailable")
+    end
+
+    if !(File.exists?(assign_dir))
+      # TODO: show the error to user instead of raising a runtime error
+      raise I18n.t("automated_tests.source_files_unavailable")
+    end
+
+    dir_contents = Dir.entries(assign_dir)
+
+    #if there are no files in repo (ie only the current and parent directory pointers)
+    if (dir_contents.length <= 2)
+      raise I18n.t("automated_tests.source_files_unavailable")
+    end
+
+    scripts = TestScript.where(assignment_id: @assignment.id)
+    if scripts.empty?
+      # TODO: show the error to user instead of raising a runtime error
+      raise I18n.t("automated_tests.test_files_unavailable")
+    end
+
+    return true
+  end
+
+  # Verify the user has the permission to run the tests - admin
+  # and graders always have the permission, while student has to
+  # belong to the group, and have at least one token.
+  def self.has_permission?()
+    if @current_user.admin?
+      true
+    elsif @current_user.ta?
+      true
+    elsif @current_user.student?
+      # Make sure student belongs to this group
+      if not @current_user.accepted_groupings.include?(@grouping)
+        # TODO: show the error to user instead of raising a runtime error
+        raise I18n.t("automated_tests.not_belong_to_group")
+      end
+      #can skip checking tokens if we have unlimited
+      if @grouping.assignment.unlimited_tokens
+        return true
+      end
+      t = @grouping.token
+      if t == nil
+        raise I18n.t('automated_tests.missing_tokens')
+      end
+      if t.tokens > 0
+        t.decrease_tokens
+        true
+      else
+        # TODO: show the error to user instead of raising a runtime error
+        raise I18n.t("automated_tests.missing_tokens")
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.